src/Features/CSharp/Portable/CodeRefactorings/UseExplicitType/UseExplicitTypeCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,139 @@
+    {
+        private static readonly CSharpUseExplicitTypeHelper s_useExplicitTypeHelper = new CSharpUseExplicitTypeHelper();
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var textSpan = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (!textSpan.IsEmpty)
+            {
+                return;
+            }
+
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            var declaration = GetDeclaration(root, textSpan);
+            Debug.Assert(declaration.IsKind(SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement, SyntaxKind.DeclarationExpression));
+
+            var declaredType = s_useExplicitTypeHelper.FindAnalyzableType(declaration, semanticModel, cancellationToken);
+            if (declaredType == null)
+            {
+                return;
+            }
+
+            if (declaredType.OverlapsHiddenPosition(cancellationToken))
+            {
+                return;
+            }
+
+            if (!declaredType.Span.IntersectsWith(textSpan.Start))
+            {
+                return;
+            }
+
+            var state = State.Generate(declaration, semanticModel, optionSet,
+                isVariableDeclarationContext: declaration.IsKind(SyntaxKind.VariableDeclaration), cancellationToken: cancellationToken);
+
+            // UseExplicitType analyzer/fixer already gives an action in this case
+            if (s_useExplicitTypeHelper.IsStylePreferred(semanticModel, optionSet, state, cancellationToken))
+            {
+                return;
+            }
+
+            Debug.Assert(state != null, "analyzing a declaration and state is null.");
+            if (!s_useExplicitTypeHelper.TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    CSharpFeaturesResources.Use_explicit_type,
+                    c => UpdateDocumentAsync(document, declaredType, c)));
+        }
+
+        private static SyntaxNode GetDeclaration(SyntaxNode root, TextSpan textSpan)
+        {
+            // Ex:
+            // var (x, y) = ...
+            // (var x, ...) = ...
+            // foreach (var (x, y) in ...) ...
+            var declarationExpression = root.FindToken(textSpan.Start).GetAncestor<DeclarationExpressionSyntax>();
```

Comment:
1. you call root.FindToken 3 times with teh same arg.  Considre just doing it once.
2. you coudl simplify with ```token.GetAncestor<VarDecl>() ?? token.GetAncestor<...>() ?? token.GetAncestor<...>()```.
3. this logic looks wrong.  esp the log about variable declaration and foreach.  Specifically (due to lambdas) you could have these nodes contained within each other.  What you're really trying to say is "find the closest containing ancestor of any of these types".  However, this code is saying "find the closest variable, then find the closest foreach" with the presumption that the closest variable is closer than the closest foreach.

For '3' you sholud be able to make a test whre you're inside a foreach-variable inside a lambda inside var-decl.  The feature will fail in this case i believe.

Reply:
Thanks a lot. The logic was definitely wrong.

---
In reply to: [172418520](https://github.com/dotnet/roslyn/pull/25208#discussion_r172418520) [](ancestors = 172418520)